### PR TITLE
Fix nonzero of a SparseArray

### LINKED ISSUE #21172: Method dropna does not work on SparseDataFrames
Function `dropna` may return wrong result on `SparseDataFrame`.  The following code

```python
import pandas as pd

pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [float('nan'), 0]}).dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [None, 0]}).dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1], "F3": [float('nan'), 0]}).dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1]}).dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1]}).dropna(axis=1, inplace=False, how='all')

pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [float('nan'), 0]}).to_dense().dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [None, 0]}).to_dense().dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1], "F3": [float('nan'), 0]}).to_dense().dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1]}).to_dense().dropna(axis=1, inplace=False, how='all')
pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1]}).to_dense().dropna(axis=1, inplace=False, how='all')
```

outputs 

```python
import pandas as pd

print(pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [float('nan'), 0]}).dropna(axis=1, inplace=False, how='all'))
    F1  F2
0  NaN   0
1  NaN   1

print(pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [None, 0]}).dropna(axis=1, inplace=False, how='all'))
    F1  F2
0  NaN   0
1  NaN   1

print(pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1], "F3": [float('nan'), 0]}).dropna(axis=1, inplace=False, how='all'))
   F1  F2
0 NaN   0
1 NaN   1

print(pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1]}).dropna(axis=1, inplace=False, how='all'))
    F1
0  NaN
1  NaN

print(pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1]}).dropna(axis=1, inplace=False, how='all'))
   F1
0 NaN
1 NaN

print(pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [float('nan'), 0]}).to_dense().dropna(axis=1, inplace=False, how='all'))
   F2   F3
0   0  NaN
1   1  0.0

print(pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1], "F3": [None, 0]}).to_dense().dropna(axis=1, inplace=False, how='all'))
   F2   F3
0   0  NaN
1   1  0.0

print(pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1], "F3": [float('nan'), 0]}).to_dense().dropna(axis=1, inplace=False, how='all'))
   F2   F3
0   0  NaN
1   1  0.0

print(pd.SparseDataFrame({"F1": [None, None], "F2": [0, 1]}).to_dense().dropna(axis=1, inplace=False, how='all'))
   F2
0   0
1   1

print(pd.SparseDataFrame({"F1": [float('nan'), float('nan')], "F2": [0, 1]}).to_dense().dropna(axis=1, inplace=False, how='all'))
   F2
0   0
1   1
```

#### Problem description

`dropna` method behaves differently for `SparseDataFrame`s and dense ones. Also it may happen that it does not drop `nan` columns at all (see the last examples in the first batch). The correct behaviour is in the second batch of commands.

#### Expected Output

```
   F2   F3
0   0  NaN
1   1  0.0

   F2   F3
0   0  NaN
1   1  0.0

   F2   F3
0   0  NaN
1   1  0.0

   F2
0   0
1   1

   F2
0   0
1   1
```

<details>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64  
OS: Linux       
OS-release: 4.15.0-20-generic
machine: x86_64
processor:
byteorder: little                                                                
LC_ALL: C.UTF-8
LANG: C.UTF-8
LOCALE: en_US.UTF-8
                                                                                 
pandas: 0.23.0
pytest: 3.5.0
pip: 9.0.3
setuptools: 39.0.1
Cython: 0.28.2                                                                   
numpy: 1.14.3
scipy: 1.0.1
pyarrow: None
xarray: None                                                                     
IPython: 6.3.1
sphinx: None
patsy: None
dateutil: 2.7.2                                                                  
pytz: 2018.4
blosc: None
bottleneck: None
tables: None                                                                     
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: None                                                                   
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1       
sqlalchemy: 1.2.7
pymysql: None     
psycopg2: None    
jinja2: 2.10
s3fs: None           
fastparquet: None
pandas_gbq: None
pandas_datareader: None
```

</details>

### PR DESCRIPTION
**Pull request by babky**
The nonzero operation returned the nonzero locations of the underlying index. However we need to get the nonzero locations in the real array. For this operation to be faster an inverse index structure would be beneficial or it could be implemented using binary search.

```python
sa = pd.SparseArray([float('nan'), float('nan'), 1, 0, 0, 2, 0, 0, 0, 3, 0, 0])
```
returned `0, 3, 7`. The index is shifted by two because of the two first `NaN`s and that's why the `0, 3, 7` are returned. The correct result would be `2, 5, 9` and is found in the method.

For the above sample the code works. However for other implementations of `SparseIndex` it could be broken.

- [x] closes #21172
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
@babky int32 isn’t imported, needs to be changed to np.int32.

**Comment by jbrockmendel** [comment]
@babky small fixup needed to get this working

**Comment by babky** [comment]
hi, i will fix them in near future, was on a vacation...

**Comment by jreback** [comment]
needs tests, whatsnew entry, and should be a vectorized soln.

**Comment by jorisvandenbossche** [comment]
cc @TomAugspurger pinging you here since you are working on sparse, to make sure this PR would not conflict with your work

**Comment by TomAugspurger** [comment]
At a glance, this approach won’t work with my implementation since calls super to get ndarray.nonzero

**Comment by babky** [comment]
The issue looks gone in the current master. I've added some unit tests to cover the issue. From my really fast overlook of the tests I think that the tests could be added.

**Comment by babky** [comment]
@jbrockmendel @TomAugspurger: I believe the issue is gone in master branch and I just added the tests which should prevent a regression.

File: pandas/tests/sparse/test_indexing.py [hunk 1]
```diff
@@ -455,6 +455,25 @@ def tests_indexing_with_sparse(self, kind, fill):
             s.iloc[indexer]
 
 
+class TestSparseArray(object):
```
Comment (TomAugspurger) [inline pandas/tests/sparse/test_indexing.py]: Could you move this to `pandas/tests/arrays/sparse/test_array.py`?

**Review by TomAugspurger — COMMENTED** [review]
Can you also add a whatsnew in 0.24.0.txt? I don't see one saying that we fixed nonzero.

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Moved it one more time. We reorganized things.

File: doc/source/whatsnew/v0.24.0.txt [hunk 2]
```diff
@@ -1353,6 +1353,7 @@ Sparse
 - Bug in ``DataFrame.groupby`` not including ``fill_value`` in the groups for non-NA ``fill_value`` when grouping by a sparse column (:issue:`5078`)
 - Bug in unary inversion operator (``~``) on a ``SparseSeries`` with boolean values. The performance of this has also been improved (:issue:`22835`)
 - Bug in :meth:`SparseArary.unique` not returning the unique values (:issue:`19595`)
+- Bug in ``SparseArray.nonzero`` and `SparseDataFrame.dropna` returning shifted/invalid results (:issue:`21172`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you add ``:func:`` to reference the api of these

File: pandas/tests/arrays/sparse/test_array.py [hunk 3]
```diff
@@ -747,6 +747,22 @@ def test_fillna_overlap(self):
         exp = SparseArray([1, 3, 3, 3, 3], fill_value=0, dtype=np.float64)
         tm.assert_sp_array_equal(res, exp)
 
+    def test_nonzero(self):
+        sa = pd.SparseArray([
```
Comment (jreback) [inline pandas/tests/arrays/sparse/test_array.py]: can you add the issue number

File: pandas/tests/arrays/sparse/test_array.py [hunk 4]
```diff
@@ -747,6 +747,22 @@ def test_fillna_overlap(self):
...
+            float('nan'),
+            float('nan'),
+            1, 0, 0,
+            2, 0, 0, 0,
+            3, 0, 0
+        ])
+        tm.assert_numpy_array_equal(np.array([2, 5, 9], dtype=np.int32),
```
Comment (jreback) [inline pandas/tests/arrays/sparse/test_array.py]: can you use 

```
result = 
expected = 
tm.assert_numpy_array_equal(result, expected)
```

File: pandas/tests/sparse/frame/test_frame.py [hunk 5]
```diff
@@ -1350,3 +1350,11 @@ def test_assign_with_sparse_frame(self):
 
         for column in res.columns:
             assert type(res[column]) is SparseSeries
+
+    def test_dropna(self):
+        tm.assert_sp_frame_equal(
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: use the result= and expected= format

add the issue number as a comment

File: pandas/tests/sparse/frame/test_frame.py [hunk 6]
```diff
@@ -1350,3 +1350,11 @@ def test_assign_with_sparse_frame(self):
...
+            pd.SparseDataFrame({"F2": [0, 1]}),
+            pd.SparseDataFrame(
+                {"F1": [float('nan'), float('nan')], "F2": [0, 1]}
+            ).dropna(axis=1, inplace=False, how='all')
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: can you test for inplace=True/False, how='all', 'any'

via parametrization

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 2 above)
Comment (babky) [reply doc/source/whatsnew/v0.24.0.txt]: sure

**Comment by babky** [comment]
@jreback the comments should be resolved now

File: df_info.txt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<class 'pandas.core.frame.DataFrame'>
```
[the remark this answers is not in the document]
Comment (babky) [reply df_info.txt]: Yes :-( :+1:

File: pandas/tests/sparse/frame/test_frame.py [hunk 8]
```diff
@@ -1360,3 +1361,15 @@ def test_assign_with_sparse_frame(self):
 
         for column in res.columns:
             assert type(res[column]) is SparseSeries
+
+    def test_dropna(self):
+        # Tests regression #21172.
+        expected = pd.SparseDataFrame({"F2": [0, 1]})
+        for inplace, how in product((True, False), ('all', 'any')):
```
[the remark this answers is not in the document]
Comment (babky) [reply pandas/tests/sparse/frame/test_frame.py]: ouch, did not know that, thank you :taco:

**Review by jreback — CHANGES_REQUESTED** [review]
if you remove the df_info file and ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @babky

**Pull request merged**